pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -0,0 +1,478 @@
+
+	<-stopCh
+}
+
+func (tc *Controller) addJob(obj interface{}) {
+	job := obj.(*batch.Job)
+	glog.V(4).Infof("Adding job %s/%s", job.Namespace, job.Name)
+
+	if job.Spec.TTLSecondsAfterFinished != nil {
+		tc.enqueue(job)
+	}
+}
+
+func (tc *Controller) updateJob(old, cur interface{}) {
+	job := cur.(*batch.Job)
+	glog.V(4).Infof("Updating job %s/%s", job.Namespace, job.Name)
+
+	if job.Spec.TTLSecondsAfterFinished != nil {
+		tc.enqueue(job)
+	}
+}
+
+func (tc *Controller) deleteJob(obj interface{}) {
+	job, ok := obj.(*batch.Job)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		job, ok = tombstone.Obj.(*batch.Job)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a job %#v", obj))
+			return
+		}
+	}
+	glog.V(4).Infof("Job %s/%s deleted", job.Namespace, job.Name)
+	item, err := itemFromJob(job)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	tc.jobsDelayCleanupQueue.Delete(item)
+}
+
+func (tc *Controller) enqueue(job *batch.Job) {
+	glog.V(4).Infof("Add job %s/%s to cleanup", job.Namespace, job.Name)
+	key, err := controller.KeyFunc(job)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", job, err))
+		return
+	}
+
+	tc.queue.Add(key)
+}
+
+func (tc *Controller) worker() {
+	for tc.processNextWorkItem() {
+	}
+}
+
+func (tc *Controller) processNextWorkItem() bool {
+	key, quit := tc.queue.Get()
+	if quit {
+		return false
+	}
+	defer tc.queue.Done(key)
+
+	err := tc.processJob(key.(string))
+	tc.handleErr(err, key)
+
+	return true
+}
+
+func (tc *Controller) runJobDelayCleanupWorker() {
+	for tc.handleJobsDelayCleanup() {
+	}
+}
+
+func (tc *Controller) handleJobsDelayCleanup() bool {
+	if tc.jobsDelayCleanupQueue.IsClosed() {
+		glog.V(2).Infof("The delay cleanup queue for Jobs is closed.")
+		return false
+	}
+
+	nowUTC := tc.clock.Now().UTC()
+
+	tc.nextCheckAtMutex.RLock()
+	checkAt := tc.nextCheckAt
+	tc.nextCheckAtMutex.RUnlock()
+
+	// No need to check delay cleanup queue for now.
+	if checkAt == nil || checkAt.IsZero() || nowUTC.Before(checkAt.UTC()) {
+		return true
+	}
+
+	tc.jobsDelayCleanupQueueMutex.Lock()
+	defer tc.jobsDelayCleanupQueueMutex.Unlock()
+
+	glog.V(4).Infof("Handling Jobs in the delay cleanup queue")
+	// Keep finding expired Jobs from the delay cleanup queue, and send them to clean up queue.
+	for {
+		obj := tc.jobsDelayCleanupQueue.Peek()
+		if obj == nil {
+			// Nothing in the queue.
+			// Clear the next check time.
+			tc.clearNextCheckAt()
+			return true
+		}
+		top, ok := obj.(*Item)
+		if !ok || top == nil {
+			glog.Errorf("Error peeking object from the delay cleanup queue, it's not a valid type, dropping it: %#v", obj)
+			tc.popJobsDelayCleanupQueue(top)
+			continue
+		}
+		if top.ttl == nil {
```

Comment:
shouldn't you get the value present value from an informer? I'd not store it in the Item struct as it can can more outdated